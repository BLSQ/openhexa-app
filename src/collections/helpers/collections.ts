import { gql } from "@apollo/client";
import { getApolloClient } from "core/helpers/apollo";
import { CreateCollectionInput } from "graphql-types";
import {
  CreateCollectionMutationVariables,
  CreateCollectionMutation,
  AddDhis2DataElementToCollectionMutation,
  AddDhis2DataElementToCollectionMutationVariables,
  AddS3ObjectToCollectionMutation,
  AddS3ObjectToCollectionMutationVariables,
} from "./collections.generated";

export async function addToCollection(
  collectionId: string,
  element: { id: string; __typename: string }
) {
  const client = getApolloClient();
  switch (element.__typename) {
    case "DHIS2DataElement": {
      const { data } = await client.mutate<
        AddDhis2DataElementToCollectionMutation,
        AddDhis2DataElementToCollectionMutationVariables
      >({
        mutation: gql`
          mutation addDHIS2DataElementToCollection(
            $input: AddDHIS2DataElementToCollectionInput!
          ) {
            addToCollection: addDHIS2DataElementToCollection(input: $input) {
              success
              errors
            }
          }
        `,
        variables: { input: { id: element.id, collectionId } },
      });
      return data?.addToCollection.success;
    }
    case "S3Object": {
      const { data } = await client.mutate<
        AddS3ObjectToCollectionMutation,
        AddS3ObjectToCollectionMutationVariables
      >({
        mutation: gql`
          mutation addS3ObjectToCollection(
            $input: AddS3ObjectToCollectionInput!
          ) {
            addToCollection: addS3ObjectToCollection(input: $input) {
              success
              errors
            }
          }
        `,
        variables: { input: { id: element.id, collectionId } },
      });
      return data?.addToCollection.success;
    }
  }
}

export async function removeFromCollection(
  collectionId: string,
  element: { id: string; __typename: string }
) {
  const client = getApolloClient();
  switch (element.__typename) {
    case "DHIS2DataElement": {
      const { data } = await client.mutate({
        mutation: gql`
          mutation removeDHIS2DataElementFromCollection(
            $input: RemoveDHIS2DataElementFromCollectionInput!
          ) {
            removeFromCollection: removeDHIS2DataElementFromCollection(
              input: $input
            ) {
              success
              errors
            }
          }
        `,
        variables: {
          input: {
            collectionId,
            id: element.id,
          },
        },
      });
      return data?.removeFromCollection.success;
    }
    case "S3Object": {
      const { data } = await client.mutate({
        mutation: gql`
          mutation removeS3ObjectFromCollection(
            $input: RemoveS3ObjectFromCollectionInput!
          ) {
            removeFromCollection: removeS3ObjectFromCollection(input: $input) {
              success
              errors
            }
          }
        `,
        variables: {
          input: {
            collectionId,
            id: element.id,
          },
        },
      });
      return data?.removeFromCollection.success;
    }
  }
}

export async function createCollection(input: CreateCollectionInput) {
  const client = getApolloClient();
  const { data } = await client.mutate<
    CreateCollectionMutation,
    CreateCollectionMutationVariables
  >({
    mutation: gql`
      mutation CreateCollection($input: CreateCollectionInput!) {
        createCollection(input: $input) {
          success
          errors
          collection {
            id
            name
          }
        }
      }
    `,
    variables: { input },
  });

  if (!data?.createCollection.collection) {
    throw new Error("Unable to create collection.");
  }

  return data?.createCollection.collection;
}
