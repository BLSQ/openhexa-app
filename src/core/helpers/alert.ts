export enum AlertType {
  success,
  error,
  info,
  warning,
}

export function displayAlert(message: string, type?: AlertType) {
  if (!window) {
    console.warn("displayAlert cannot be called on the server.");
    return;
  }
  const event = new CustomEvent("displayAlert", {
    detail: {
      message,
      type: type ?? AlertType.info,
    },
  });

  window.dispatchEvent(event);
}
