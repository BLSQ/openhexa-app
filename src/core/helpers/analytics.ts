import { NextRequest } from "next/server";
import getConfig from "next/config";

const { publicRuntimeConfig } = getConfig();

interface TrackEventProperties {
  [key: string]: any;
}

async function sendEvent(
  event: string,
  properties: TrackEventProperties,
  headers?: Headers,
): Promise<void> {
  if (publicRuntimeConfig.DISABLE_ANALYTICS === "true") {
    return;
  }
  const res = await fetch(
    `${publicRuntimeConfig.OPENHEXA_BACKEND_URL ?? ""}/analytics/track/`,
    {
      method: "POST",
      priority: "low",
      headers: headers,
      credentials: "include",
      body: JSON.stringify({ event, properties }),
    },
  );

  if (!res.ok) {
    throw new Error(`Failed to send event: ${res.statusText}`);
  }
}

export async function trackEvent(
  event: string,
  properties: TrackEventProperties,
): Promise<void> {
  if (typeof window === "undefined") {
    // We do not support tracking events on the server
    return;
  }
  try {
    await sendEvent(event, properties);
  } catch (error) {
    console.error("Error tracking event:", error);
  }
}

export async function pageView(request: NextRequest) {
  // This is to be called on the server side only.
  if (typeof window !== "undefined") {
    return;
  }
  try {
    const headers = request.headers;
    headers.set("content-type", "application/json");

    await sendEvent(
      "pageview",
      {
        pathname: request.nextUrl.pathname,
        searchParams: request.nextUrl.searchParams.toString(),
      },
      headers,
    );
  } catch (error) {
    console.error("Error tracking event:", error);
  }
}
