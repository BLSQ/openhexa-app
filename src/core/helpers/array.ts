export function ensureArray<T>(maybeArray: T[] | T): T[] {
  if (typeof maybeArray === "undefined") {
    return [];
  } else if (maybeArray === null) {
    return [];
  } else if (Array.isArray(maybeArray)) {
    return maybeArray;
  } else {
    return new Array(maybeArray);
  }
}
