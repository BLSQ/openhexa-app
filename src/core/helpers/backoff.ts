export type BackoffOptions = {
  min: number;
  max: number;
  factor: number;
  jitter: number;
};

class Backoff {
  private _attempts: number;
  private _base: number;
  private _max: number;
  private _factor: number;
  private _jitter: number;

  constructor(options: BackoffOptions) {
    this._base = options.min || 100;
    this._max = options.max || 10000;
    this._factor = options.factor || 2;
    this._jitter =
      options.jitter > 0 && options.jitter < 1 ? options.jitter : 0;
    this._attempts = 0;
  }

  get attempts(): number {
    return this._attempts;
  }

  duration() {
    let duration = this._base * Math.pow(this._factor, this._attempts++);
    if (this._jitter) {
      var rand = Math.random();
      var deviation = Math.floor(rand * this._jitter * duration);
      duration =
        (Math.floor(rand * 10) & 1) == 0
          ? duration - deviation
          : duration + deviation;
    }
    return Math.min(duration, this._max) | 0;
  }

  reset() {
    this._attempts = 0;
  }
}

export default Backoff;
