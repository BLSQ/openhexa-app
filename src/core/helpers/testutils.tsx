import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { faker } from "@faker-js/faker";
import { ByRoleOptions, screen } from "@testing-library/react";
import { ReactNode } from "react";

type TestAppProps = {
  children: ReactNode;
  mocks?: MockedResponse[];
};

jest.mock("next/router", () => require("next-router-mock"));
// This is needed for mocking 'next/link':
jest.mock("next/dist/client/router", () => require("next-router-mock"));
jest.mock("identity/hooks/useMe");

export function TestApp(props: TestAppProps) {
  return (
    <MockedProvider addTypename={false} mocks={props.mocks ?? []}>
      {props.children}
    </MockedProvider>
  );
}

export const LOGGED_IN_USER = {
  id: faker.datatype.uuid(),
  firstName: faker.name.firstName,
  lastName: faker.name.lastName,
  avatar: {
    initials: "AE",
    color: "red",
  },
};

export async function waitForDialog(options?: ByRoleOptions) {
  const dialog = await screen.getByRole("dialog", options);
  return dialog;
}
