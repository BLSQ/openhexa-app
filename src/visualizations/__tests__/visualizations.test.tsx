import { render, screen, waitFor } from "@testing-library/react";
import { DateTime, Settings } from "luxon";
import { TestApp } from "core/helpers/testutils";
import VisualizationsPage from "pages/visualizations";
import VisualizationPage from "pages/visualizations/[visualizationId]";
import {
  VisualizationDocument,
  VisualizationsPageDocument,
} from "visualizations/graphql/queries.generated";

describe("Visualization", () => {
  it("renders the visualizations' page", async () => {
    const graphqlMocks = [
      {
        request: {
          query: VisualizationsPageDocument,
          variables: {
            page: 1,
            perPage: 1,
          },
        },
        result: {
          data: {
            externalDashboards: {
              totalPages: 0,
              totalItems: 0,
              items: [],
            },
          },
        },
      },
    ];

    const { container } = render(
      <TestApp mocks={graphqlMocks}>
        <VisualizationsPage page={1} perPage={1} />
      </TestApp>,
    );
    const elm = await screen.findByText("Visualizations", { selector: "h2" });
    expect(elm).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("renders the visualizations' page with data", async () => {
    (Settings.now as jest.Mock).mockReturnValue(
      DateTime.fromObject({ year: 2022, month: 10, day: 22 }).toMillis(),
    );

    const visualizationMock = [
      {
        request: {
          query: VisualizationsPageDocument,
          variables: {
            page: 1,
            perPage: 1,
          },
        },
        result: {
          data: {
            externalDashboards: {
              totalPages: 1,
              totalItems: 1,
              items: [
                {
                  id: "351489ac-5bbe-4ed3-aab9-8155024092b4",
                  name: "Untitled Dashboard",
                  url: "https://coronavirus.data.gov.uk/",
                  pictureUrl:
                    "/visualizations/351489ac-5bbe-4ed3-aab9-8155024092b4/image/",
                  createdAt: "2022-10-21T11:35:42.716Z",
                  updatedAt: "2022-10-21T11:35:42.716Z",
                  countries: [],
                  description: "",
                  tags: [],
                },
              ],
            },
          },
        },
      },
    ];
    const { container } = render(
      <TestApp mocks={visualizationMock}>
        <VisualizationsPage page={1} perPage={1} />
      </TestApp>,
    );

    const elm = await screen.findByText("Visualizations", { selector: "h2" });
    expect(elm).toBeInTheDocument();
    expect(container).toMatchSnapshot();

    expect(await screen.findByText("Untitled Dashboard")).toBeInTheDocument();
  });

  it("renders a visualization info page ", async () => {
    const visualizationMock = [
      {
        request: {
          query: VisualizationDocument,
          variables: {
            id: "351489ac-5bbe-4ed3-aab9-8155024092b4",
          },
        },
        result: {
          data: {
            externalDashboard: {
              id: "351489ac-5bbe-4ed3-aab9-8155024092b4",
              name: "Untitled Dashboard",
              url: "https://coronavirus.data.gov.uk/",
              pictureUrl:
                "/visualizations/351489ac-5bbe-4ed3-aab9-8155024092b4/image/",
              createdAt: "2022-10-21T11:35:42.716Z",
              updatedAt: "2022-10-21T11:35:42.716Z",
              countries: [],
              description: "",
              tags: [],
            },
          },
        },
      },
    ];

    const { container } = render(
      <TestApp mocks={visualizationMock}>
        <VisualizationPage visualizationId="351489ac-5bbe-4ed3-aab9-8155024092b4" />
      </TestApp>,
    );

    const elm = await screen.findByText("https://coronavirus.data.gov.uk/");
    expect(elm).toBeInTheDocument();

    waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it("renders null as the visualization does not exist", async () => {
    const visualizationMock = [
      {
        request: {
          query: VisualizationDocument,
          variables: {
            id: "151489ac-5bbe-4ed3-aab9-8155024092b4",
          },
        },
        result: {
          data: {
            externalDashboard: null,
          },
        },
      },
    ];

    const { container } = render(
      <TestApp mocks={visualizationMock}>
        <VisualizationPage visualizationId="151489ac-5bbe-4ed3-aab9-8155024092b4" />
      </TestApp>,
    );

    expect(container.firstChild).toBeNull();
    expect(container).toMatchSnapshot();
  });
});
