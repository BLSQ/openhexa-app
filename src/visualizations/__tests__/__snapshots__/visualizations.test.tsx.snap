// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualization renders a visualization info page  1`] = `
<div>
  <main
    class="w-full max-w-7xl mx-auto px-2 sm:px-6 lg:px-8 pb-6"
  >
    <nav
      aria-label="Breadcrumbs"
      class="my-8 px-2"
    >
      <ol
        class="flex items-center"
        role="list"
      >
        <li>
          <div>
            <a
              class="text-gray-400 hover:text-gray-500"
              href="/"
            >
              <svg
                aria-hidden="true"
                class="h-5 w-5 flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M9.293 2.293a1 1 0 011.414 0l7 7A1 1 0 0117 11h-1v6a1 1 0 01-1 1h-2a1 1 0 01-1-1v-3a1 1 0 00-1-1H9a1 1 0 00-1 1v3a1 1 0 01-1 1H5a1 1 0 01-1-1v-6H3a1 1 0 01-.707-1.707l7-7z"
                  fill-rule="evenodd"
                />
              </svg>
              <span
                class="sr-only"
              >
                Home
              </span>
            </a>
          </div>
        </li>
        <li
          class="ml-2"
        >
          <div
            class="flex items-center"
          >
            <svg
              aria-hidden="true"
              class="h-5 w-5 flex-shrink-0 text-gray-300"
              fill="currentColor"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.555 17.776l8-16 .894.448-8 16-.894-.448z"
              />
            </svg>
            <span
              class="ml-3 text-sm hover:text-gray-700 font-medium text-gray-500"
            >
              <a
                class=""
                href="/visualizations"
              >
                Visualizations
              </a>
            </span>
          </div>
        </li>
        <li
          class="ml-2"
        >
          <div
            class="flex items-center"
          >
            <svg
              aria-hidden="true"
              class="h-5 w-5 flex-shrink-0 text-gray-300"
              fill="currentColor"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.555 17.776l8-16 .894.448-8 16-.894-.448z"
              />
            </svg>
            <span
              class="ml-3 text-sm hover:text-gray-700 font-medium text-gray-500"
            >
              <a
                class=""
                href="/visualizations/351489ac-5bbe-4ed3-aab9-8155024092b4"
              >
                Untitled Dashboard
              </a>
            </span>
          </div>
        </li>
      </ol>
    </nav>
    <div
      class="space-y-10"
    >
      <article
        class="sm:rounded-lg overflow-hidden shadow border-b border-gray-200 bg-white divide-y divide-gray-200"
      >
        <h3
          class="px-4 py-5 pt-7 text-lg font-medium leading-4 text-gray-900 sm:px-6 flex items-center justify-between"
        >
          <div>
            <div
              class="flex items-center"
            >
              <div
                class="flex h-10 w-10 flex-shrink-0 items-center"
              >
                <img
                  alt=""
                  class="h-10 w-10 rounded-full"
                />
              </div>
              <div
                class="ml-4 w-full truncate"
              >
                <div
                  class="truncate text-sm font-medium text-gray-900"
                  title="Untitled Dashboard"
                >
                  Untitled Dashboard
                </div>
                <div
                  class="truncate text-sm text-gray-500"
                >
                  <span>
                    External Dashboard
                  </span>
                </div>
              </div>
            </div>
          </div>
        </h3>
        <section
          class="px-4 py-5 sm:px-6"
        >
          <div>
            <div
              aria-expanded="true"
              class="-my-7 flex items-center py-7"
              data-headlessui-state="open"
              id="headlessui-disclosure-button-6fbe024f-2316-4265-a6e8-d65a837e308a"
            >
              <h4
                class="font-medium"
              >
                Information
              </h4>
              <div
                class="flex flex-1 flex-shrink items-center justify-end"
              />
            </div>
            <div>
              <div
                class="mt-6"
                data-headlessui-state="open"
                id="headlessui-disclosure-panel-e67862f3-ccbf-4c51-b8ed-ed1d0420ea19"
              >
                <dl
                  class="grid gap-6"
                >
                  <div
                    class="grid grid-cols-5 gap-2 sm:gap-4"
                  >
                    <dt
                      class="flex text-sm font-medium text-gray-500 col-span-1"
                    >
                      <span
                        class="inline"
                      >
                        Url
                      </span>
                    </dt>
                    <dd
                      class="col-span-4 text-sm text-gray-900"
                    >
                      <a
                        class="flex items-center text-blue-600 hover:text-blue-500 focus:outline-none"
                        href="https://coronavirus.data.gov.uk/"
                      >
                        https://coronavirus.data.gov.uk/
                      </a>
                    </dd>
                  </div>
                </dl>
              </div>
            </div>
          </div>
        </section>
        <section
          class="px-4 py-5 sm:px-6"
        >
          <div>
            <div
              aria-expanded="true"
              class="-my-7 flex items-center py-7"
              data-headlessui-state="open"
              id="headlessui-disclosure-button-6af683b4-52ac-4d60-8c9f-ab447c31177e"
            >
              <h4
                class="font-medium"
              >
                OpenHexa Metadata
              </h4>
              <div>
                <button
                  class="ml-4 inline-flex items-center gap-1 text-sm text-blue-500 hover:text-blue-400"
                >
                  Edit
                  <svg
                    aria-hidden="true"
                    class="h-4"
                    fill="none"
                    stroke="currentColor"
                    stroke-width="1.5"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M16.862 4.487l1.687-1.688a1.875 1.875 0 112.652 2.652L6.832 19.82a4.5 4.5 0 01-1.897 1.13l-2.685.8.8-2.685a4.5 4.5 0 011.13-1.897L16.863 4.487zm0 0L19.5 7.125"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="flex flex-1 flex-shrink items-center justify-end"
              />
            </div>
            <div>
              <div
                class="mt-6"
                data-headlessui-state="open"
                id="headlessui-disclosure-panel-14e41280-1ae3-4e44-b909-992f94b3186f"
              >
                <dl
                  class="grid gap-6"
                >
                  <div
                    class="grid grid-cols-5 gap-2 sm:gap-4"
                  >
                    <dt
                      class="flex text-sm font-medium text-gray-500 col-span-1"
                    >
                      <span
                        class="inline"
                      >
                        Label
                      </span>
                    </dt>
                    <dd
                      class="col-span-4 text-sm text-gray-900"
                    >
                      <div
                        class="prose text-sm text-gray-900"
                      >
                        Untitled Dashboard
                      </div>
                    </dd>
                  </div>
                  <div
                    class="grid grid-cols-5 gap-2 sm:gap-4"
                  >
                    <dt
                      class="flex text-sm font-medium text-gray-500 col-span-1"
                    >
                      <span
                        class="inline"
                      >
                        Description
                      </span>
                    </dt>
                    <dd
                      class="col-span-4 text-sm text-gray-900"
                    >
                      <div
                        class="prose text-sm text-gray-900"
                      />
                    </dd>
                  </div>
                  <div
                    class="grid grid-cols-5 gap-2 sm:gap-4"
                  >
                    <dt
                      class="flex text-sm font-medium text-gray-500 col-span-1"
                    >
                      <span
                        class="inline"
                      >
                        Owner
                      </span>
                    </dt>
                    <dd
                      class="col-span-4 text-sm text-gray-900"
                    >
                      -
                    </dd>
                  </div>
                  <div
                    class="grid grid-cols-5 gap-2 sm:gap-4"
                  >
                    <dt
                      class="flex text-sm font-medium text-gray-500 col-span-1"
                    >
                      <span
                        class="inline"
                      >
                        Tags
                      </span>
                    </dt>
                    <dd
                      class="col-span-4 text-sm text-gray-900"
                    >
                      <div
                        class="flex flex-wrap items-center gap-1.5"
                      >
                        -
                      </div>
                    </dd>
                  </div>
                  <div
                    class="grid grid-cols-5 gap-2 sm:gap-4"
                  >
                    <dt
                      class="flex text-sm font-medium text-gray-500 col-span-1"
                    >
                      <span
                        class="inline"
                      >
                        Location
                      </span>
                    </dt>
                    <dd
                      class="col-span-4 text-sm text-gray-900"
                    >
                      <div
                        class="flex flex-wrap items-center gap-1.5"
                      >
                        -
                      </div>
                    </dd>
                  </div>
                </dl>
              </div>
            </div>
          </div>
        </section>
      </article>
    </div>
  </main>
</div>
`;

exports[`Visualization renders null as the visualization does not exist 1`] = `<div />`;

exports[`Visualization renders the visualizations' page 1`] = `
<div>
  <main
    class="w-full max-w-7xl mx-auto px-2 sm:px-6 lg:px-8 pb-6"
  >
    <nav
      aria-label="Breadcrumbs"
      class="my-8 px-2"
    >
      <ol
        class="flex items-center"
        role="list"
      >
        <li>
          <div>
            <a
              class="text-gray-400 hover:text-gray-500"
              href="/"
            >
              <svg
                aria-hidden="true"
                class="h-5 w-5 flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M9.293 2.293a1 1 0 011.414 0l7 7A1 1 0 0117 11h-1v6a1 1 0 01-1 1h-2a1 1 0 01-1-1v-3a1 1 0 00-1-1H9a1 1 0 00-1 1v3a1 1 0 01-1 1H5a1 1 0 01-1-1v-6H3a1 1 0 01-.707-1.707l7-7z"
                  fill-rule="evenodd"
                />
              </svg>
              <span
                class="sr-only"
              >
                Home
              </span>
            </a>
          </div>
        </li>
        <li
          class="ml-2"
        >
          <div
            class="flex items-center"
          >
            <svg
              aria-hidden="true"
              class="h-5 w-5 flex-shrink-0 text-gray-300"
              fill="currentColor"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.555 17.776l8-16 .894.448-8 16-.894-.448z"
              />
            </svg>
            <span
              class="ml-3 text-sm hover:text-gray-700 font-medium text-gray-500"
            >
              <a
                class=""
                href="/visualizations"
              >
                Visualizations
              </a>
            </span>
          </div>
        </li>
      </ol>
    </nav>
    <div
      class="space-y-4"
    >
      <div
        class="flex items-center justify-between"
      >
        <h2
          class="text-xl font-medium"
        >
          Visualizations
        </h2>
      </div>
      <article
        class="sm:rounded-lg overflow-hidden shadow border-b border-gray-200 bg-white"
      >
        <div>
          <div
            class="relative"
          >
            <div
              class="h-full overflow-x-auto"
            >
              <div
                class="pointer-events-none absolute bottom-0 left-0 top-0 bg-gradient-to-r to-transparent transition-opacity duration-100 from-white w-12 opacity-0"
              />
              <div
                id="child"
              >
                <table
                  class="table-fixed w-full divide-y divide-gray-200 "
                  role="table"
                  style="min-width: 330px;"
                >
                  <thead
                    class=""
                  >
                    <tr
                      class=""
                      role="row"
                      style="display: flex; flex: 1 0 auto; min-width: 330px;"
                    >
                      <th
                        class="px-4 py-3 bg-gray-50 text-left text-xs font-medium text-gray-500 uppercase tracking-wider first:pl-6 last:pr-6 whitespace-nowrap"
                        colspan="1"
                        role="columnheader"
                        scope="col"
                        style="box-sizing: border-box; flex: 240 0 auto; min-width: 240px; width: 240px;"
                      >
                        Name
                      </th>
                      <th
                        class="px-4 py-3 bg-gray-50 text-left text-xs font-medium text-gray-500 uppercase tracking-wider first:pl-6 last:pr-6 whitespace-nowrap"
                        colspan="1"
                        role="columnheader"
                        scope="col"
                        style="box-sizing: border-box; flex: 150 0 auto; min-width: 30px; width: 150px;"
                      >
                        Location
                      </th>
                      <th
                        class="px-4 py-3 bg-gray-50 text-left text-xs font-medium text-gray-500 uppercase tracking-wider first:pl-6 last:pr-6 whitespace-nowrap"
                        colspan="1"
                        role="columnheader"
                        scope="col"
                        style="box-sizing: border-box; flex: 150 0 auto; min-width: 30px; width: 150px;"
                      >
                        Updated at
                      </th>
                      <th
                        class="px-4 py-3 bg-gray-50 text-left text-xs font-medium text-gray-500 uppercase tracking-wider first:pl-6 last:pr-6 whitespace-nowrap"
                        colspan="1"
                        role="columnheader"
                        scope="col"
                        style="box-sizing: border-box; flex: 100 0 auto; min-width: 30px; width: 100px;"
                      />
                    </tr>
                  </thead>
                  <tbody
                    class="divide-y divide-gray-200"
                    role="rowgroup"
                  />
                </table>
                <div
                  class="px-6 py-4 text-center text-sm italic text-gray-500"
                >
                  No elements to display
                </div>
              </div>
              <div
                class="pointer-events-none absolute bottom-0 right-0 top-0 bg-gradient-to-l to-transparent transition-opacity duration-100 from-white w-12 opacity-0"
              />
            </div>
          </div>
        </div>
      </article>
    </div>
  </main>
</div>
`;

exports[`Visualization renders the visualizations' page with data 1`] = `
<div>
  <main
    class="w-full max-w-7xl mx-auto px-2 sm:px-6 lg:px-8 pb-6"
  >
    <nav
      aria-label="Breadcrumbs"
      class="my-8 px-2"
    >
      <ol
        class="flex items-center"
        role="list"
      >
        <li>
          <div>
            <a
              class="text-gray-400 hover:text-gray-500"
              href="/"
            >
              <svg
                aria-hidden="true"
                class="h-5 w-5 flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M9.293 2.293a1 1 0 011.414 0l7 7A1 1 0 0117 11h-1v6a1 1 0 01-1 1h-2a1 1 0 01-1-1v-3a1 1 0 00-1-1H9a1 1 0 00-1 1v3a1 1 0 01-1 1H5a1 1 0 01-1-1v-6H3a1 1 0 01-.707-1.707l7-7z"
                  fill-rule="evenodd"
                />
              </svg>
              <span
                class="sr-only"
              >
                Home
              </span>
            </a>
          </div>
        </li>
        <li
          class="ml-2"
        >
          <div
            class="flex items-center"
          >
            <svg
              aria-hidden="true"
              class="h-5 w-5 flex-shrink-0 text-gray-300"
              fill="currentColor"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.555 17.776l8-16 .894.448-8 16-.894-.448z"
              />
            </svg>
            <span
              class="ml-3 text-sm hover:text-gray-700 font-medium text-gray-500"
            >
              <a
                class=""
                href="/visualizations"
              >
                Visualizations
              </a>
            </span>
          </div>
        </li>
      </ol>
    </nav>
    <div
      class="space-y-4"
    >
      <div
        class="flex items-center justify-between"
      >
        <h2
          class="text-xl font-medium"
        >
          Visualizations
        </h2>
      </div>
      <article
        class="sm:rounded-lg overflow-hidden shadow border-b border-gray-200 bg-white"
      >
        <div>
          <div
            class="relative"
          >
            <div
              class="h-full overflow-x-auto"
            >
              <div
                class="pointer-events-none absolute bottom-0 left-0 top-0 bg-gradient-to-r to-transparent transition-opacity duration-100 from-white w-12 opacity-0"
              />
              <div
                id="child"
              >
                <table
                  class="table-fixed w-full divide-y divide-gray-200 "
                  role="table"
                  style="min-width: 330px;"
                >
                  <thead
                    class=""
                  >
                    <tr
                      class=""
                      role="row"
                      style="display: flex; flex: 1 0 auto; min-width: 330px;"
                    >
                      <th
                        class="px-4 py-3 bg-gray-50 text-left text-xs font-medium text-gray-500 uppercase tracking-wider first:pl-6 last:pr-6 whitespace-nowrap"
                        colspan="1"
                        role="columnheader"
                        scope="col"
                        style="box-sizing: border-box; flex: 240 0 auto; min-width: 240px; width: 240px;"
                      >
                        Name
                      </th>
                      <th
                        class="px-4 py-3 bg-gray-50 text-left text-xs font-medium text-gray-500 uppercase tracking-wider first:pl-6 last:pr-6 whitespace-nowrap"
                        colspan="1"
                        role="columnheader"
                        scope="col"
                        style="box-sizing: border-box; flex: 150 0 auto; min-width: 30px; width: 150px;"
                      >
                        Location
                      </th>
                      <th
                        class="px-4 py-3 bg-gray-50 text-left text-xs font-medium text-gray-500 uppercase tracking-wider first:pl-6 last:pr-6 whitespace-nowrap"
                        colspan="1"
                        role="columnheader"
                        scope="col"
                        style="box-sizing: border-box; flex: 150 0 auto; min-width: 30px; width: 150px;"
                      >
                        Updated at
                      </th>
                      <th
                        class="px-4 py-3 bg-gray-50 text-left text-xs font-medium text-gray-500 uppercase tracking-wider first:pl-6 last:pr-6 whitespace-nowrap"
                        colspan="1"
                        role="columnheader"
                        scope="col"
                        style="box-sizing: border-box; flex: 100 0 auto; min-width: 30px; width: 100px;"
                      />
                    </tr>
                  </thead>
                  <tbody
                    class="divide-y divide-gray-200"
                    role="rowgroup"
                  >
                    <tr
                      class=""
                      role="row"
                      style="display: flex; flex: 1 0 auto; min-width: 330px;"
                    >
                      <td
                        class="px-4 py-5 text-sm text-gray-500 h-full first:pl-6 last:pr-6 whitespace-nowrap"
                        role="cell"
                        style="box-sizing: border-box; flex: 240 0 auto; min-width: 240px; width: 240px;"
                      >
                        <a
                          class="font-medium text-gray-700"
                          href="https://coronavirus.data.gov.uk/"
                        >
                          <div
                            class="flex items-center"
                          >
                            <div
                              class="flex h-10 w-10 flex-shrink-0 items-center"
                            >
                              <img
                                alt=""
                                class="h-10 w-10 rounded-full"
                              />
                            </div>
                            <div
                              class="ml-4 w-full truncate"
                            >
                              <div
                                class="truncate text-sm font-medium text-gray-900"
                                title="Untitled Dashboard"
                              >
                                Untitled Dashboard
                              </div>
                              <div
                                class="truncate text-sm text-gray-500"
                              >
                                https://coronavirus.data.gov.uk/
                              </div>
                            </div>
                          </div>
                        </a>
                      </td>
                      <td
                        class="px-4 py-5 text-sm text-gray-500 h-full first:pl-6 last:pr-6 whitespace-nowrap"
                        role="cell"
                        style="box-sizing: border-box; flex: 150 0 auto; min-width: 30px; width: 150px;"
                      >
                        <div
                          class="flex max-w-full flex-wrap items-center gap-2"
                        >
                          -
                        </div>
                      </td>
                      <td
                        class="px-4 py-5 text-sm text-gray-500 h-full first:pl-6 last:pr-6 whitespace-nowrap"
                        role="cell"
                        style="box-sizing: border-box; flex: 150 0 auto; min-width: 30px; width: 150px;"
                      >
                        <time
                          class="whitespace-nowrap truncate"
                          datetime="2022-10-21T13:35:42.716+02:00"
                          title="2022-10-21T13:35:42.716+02:00"
                        >
                          10 hours ago
                        </time>
                      </td>
                      <td
                        class="px-4 py-5 text-sm text-gray-500 h-full first:pl-6 last:pr-6 whitespace-nowrap"
                        role="cell"
                        style="box-sizing: border-box; flex: 100 0 auto; min-width: 30px; width: 100px;"
                      >
                        <div
                          class="w-full"
                        >
                          <a
                            class="flex w-full cursor-pointer items-center justify-end outline-none text-blue-600 hover:text-blue-500"
                            href="/visualizations/351489ac-5bbe-4ed3-aab9-8155024092b4"
                          >
                            View
                            <svg
                              aria-hidden="true"
                              class="inline h-5"
                              fill="currentColor"
                              viewBox="0 0 20 20"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                clip-rule="evenodd"
                                d="M7.21 14.77a.75.75 0 01.02-1.06L11.168 10 7.23 6.29a.75.75 0 111.04-1.08l4.5 4.25a.75.75 0 010 1.08l-4.5 4.25a.75.75 0 01-1.06-.02z"
                                fill-rule="evenodd"
                              />
                            </svg>
                          </a>
                        </div>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
              <div
                class="pointer-events-none absolute bottom-0 right-0 top-0 bg-gradient-to-l to-transparent transition-opacity duration-100 from-white w-12 opacity-0"
              />
            </div>
          </div>
        </div>
      </article>
    </div>
  </main>
</div>
`;
